wandb/sdk/launch/agent2/controllers/base.py

```diff
@@ -0,0 +1,147 @@
+from wandb.sdk.launch.errors import LaunchError
+from wandb.sdk.launch.runner.abstract import AbstractRun
+
+from ...queue_driver.abstract import AbstractQueueDriver
+from ..controller import LaunchControllerConfig, LegacyResources
+from ..jobset import Job, JobSet
+
+
+class BaseManager(ABC):
+    """Maintains state for multiple jobs."""
+
+    queue_driver: Optional[AbstractQueueDriver] = None
+
+    def __init__(
+        self,
+        config: LaunchControllerConfig,
+        jobset: JobSet,
+        logger: logging.Logger,
+        legacy: LegacyResources,
+        max_concurrency: int,
+    ):
+        self.config = config
+        self.jobset = jobset
+        self.logger = logger
+        self.legacy = legacy
+        self.max_concurrency = max_concurrency
+
+        self.id = config["jobset_spec"].name
+        self.active_runs: Dict[str, AbstractRun] = {}
+        if self.queue_driver is None:
+            raise LaunchError(
+                "queue_driver is not set, set queue driver in subclass constructor"
+            )
+
+    async def reconcile(self) -> None:
+        assert self.queue_driver is not None
+        raw_items = list(self.jobset.jobs.values())
+
+        # Dump all raw items
+        self.logger.info(
+            f"====== Raw items ======\n{json.dumps(raw_items, indent=2)}\n======================"
+        )
+
+        owned_items = {
+            item.id: item
+            for item in raw_items
+            if item.state in ["LEASED", "CLAIMED"]
+            and item.claimed_by == self.config["agent_id"]
+        }
+
+        if len(owned_items) < self.max_concurrency:
+            # we own fewer items than our max concurrency, and there are other items waiting to be run
+            # let's lease the next item
+            next_item = await self.pop_next_item()
+            if next_item is not None:
+                owned_items[next_item.id] = next_item
+
+        # ensure all our owned items are running
+        for owned_item_id in owned_items:
+            if owned_item_id not in self.active_runs:
+                # we own this item but it's not running
+                await self.launch_item(owned_items[owned_item_id])
+
+        # TODO: validate job set clears finished runs
+        # release any items that are no longer in owned items
+        for item in self.active_runs:
+            if item not in owned_items:
+                # we don't own this item anymore
+                await self.release_item(item)
+
+    async def pop_next_item(self) -> Optional[Job]:
+        assert self.queue_driver is not None
+        next_item = await self.queue_driver.pop_from_run_queue()
+        self.logger.info(f"Leased item: {json.dumps(next_item, indent=2)}")
+        return next_item
+
+    async def launch_item(self, item: Job) -> Optional[str]:
+        run_id = await self.launch_item_task(item)
+        self.logger.info(f"Launched item got run_id: {run_id}")
+        return run_id
```

Comment:
I don't think we gain anything from this indirection, all the contents of `launch_item_task` can just go here.